libraries/chain/controller.cpp

```diff
@@ -857,7 +854,9 @@ struct building_block {
 
    assembled_block assemble_block(boost::asio::io_context& ioc,
                                   const protocol_feature_set& pfs,
-                                  const block_data_t& block_data) {
+                                  const block_data_t& block_data,
+                                  bool validating,
+                                  std::optional<qc_info_t> validating_qc_info) {
```

Comment:
should take `qc_data_t`.

Instead of two parameters `validating` and `validating_qc_info`, what about we pass a single parameter:

`std::optional<incoming_block_data>` with

```
struct incoming_block_data {
   std::optional<qc_data_t> qc_data;
};
```

That way if we need to pass more stuff from the incoming block we can add it to `incoming_block_data`. It will simplify the `finish_block() interface as well.

If you want I'm happy to make the change in a derived branch.